unit/unit_fmq/unit_fmq.c

```diff
@@ -0,0 +1,471 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE 1
+#include "test_common.h"
+
+#include "gbinder_driver.h"
+#include "gbinder_fmq_p.h"
+
+#include <stdio.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+static TestOpt test_opt;
+
+typedef struct test_fmq_data {
+    const char* name;
+    gsize item_size;
+    gsize max_num_items;
+    gint type;
+    gint flags;
+    gint fd;
+    gsize buffer_size;
+} TestFmqData;
+
+/*==========================================================================*
+ * null
+ *==========================================================================*/
+
+static const TestFmqData test_fmq_tests_null[] = {
+    { "wrong_size", 0, 8,
+        GBINDER_FMQ_TYPE_SYNC_READ_WRITE, 0, -1, 0 },
+    { "wrong_count", sizeof(guint32), 0,
+        GBINDER_FMQ_TYPE_SYNC_READ_WRITE, 0, -1, 0 },
+    { "wrong_buffer_size", sizeof(guint32), 8,
+        GBINDER_FMQ_TYPE_SYNC_READ_WRITE, 0, 1, 0 },
+};
+
+static
+void
+test_null(
+    gconstpointer test_data)
+{
+    const TestFmqData* test = test_data;
+
+    GBinderFmq* fmq = gbinder_fmq_new(
+        test->item_size,
+        test->max_num_items,
+        test->type,
+        test->flags,
+        test->fd,
+        test->buffer_size);
+
+    g_assert(!fmq);
+
+    g_assert(!gbinder_fmq_ref(fmq));
+
+    g_assert(gbinder_fmq_available_to_read(fmq) == 0);
+    g_assert(gbinder_fmq_available_to_write(fmq) == 0);
+    g_assert(gbinder_fmq_available_to_read_contiguous(fmq) == 0);
+    g_assert(gbinder_fmq_available_to_write_contiguous(fmq) == 0);
+
+    g_assert(!gbinder_fmq_begin_read(fmq, 1));
+    g_assert(!gbinder_fmq_begin_read((GBinderFmq*)0x1, 0));
+    g_assert(!gbinder_fmq_begin_write(fmq, 1));
+    g_assert(!gbinder_fmq_begin_write((GBinderFmq*)0x1, 0));
+
+    g_assert(!gbinder_fmq_end_read(fmq, 1));
+    g_assert(!gbinder_fmq_end_read((GBinderFmq*)0x1, 0));
+    g_assert(!gbinder_fmq_end_write(fmq, 1));
+    g_assert(!gbinder_fmq_end_write((GBinderFmq*)0x1, 0));
+
+    g_assert(!gbinder_fmq_read(fmq, (void*)0x1, 1));
+    g_assert(!gbinder_fmq_read((GBinderFmq*)0x1, NULL, 1));
+    g_assert(!gbinder_fmq_read((GBinderFmq*)0x1, (void*)0x1, 0));
+
+    g_assert(!gbinder_fmq_write(fmq, (const void*)0x1, 1));
+    g_assert(!gbinder_fmq_write((GBinderFmq*)0x1, NULL, 1));
+    g_assert(!gbinder_fmq_write((GBinderFmq*)0x1, (const void*)0x1, 0));
+
+    g_assert(gbinder_fmq_wait_timeout(fmq, 0, (guint32*)0x1, 0) == -EINVAL);
```

Comment:
```
unit_fmq.c: In function ‘test_null’:
unit_fmq.c:109:69: error: ‘EINVAL’ undeclared (first use in this function)
  109 |     g_assert(gbinder_fmq_wait_timeout(fmq, 0, (guint32*)0x1, 0) == -EINVAL);
```
Looks like `#include <errno.h>` is needed

Reply:
Missing include added.